File: codes/matlab/ch21/singularity_subtraction.m
function singularity_subtraction(varargin)
%% singularity_subtraction - Etude 21.3: subtract a known singular part.
% f(x) = e^x + 0.1 (1+x)^(2/3) on [-1, 1].  Direct Chebyshev expansion of
% f decays algebraically; subtracting the singular part leaves e^x, which
% has geometric decay.
%
% Author: Dr. Denys Dutykh (Khalifa University, Abu Dhabi, UAE)
% Part of "Computational Etudes: A Spectral Approach"

    dump_path = parse_args(varargin{:});
    script_dir = fileparts(mfilename('fullpath'));
    addpath(script_dir);
    cm = tricks_common(); cm.configure_style();
    out_dir = cm.output_dir(script_dir);

    csing = 0.1;
    f_full     = @(x) exp(x) + csing * (1.0 + x).^(2/3);
    f_singular = @(x) csing * (1.0 + x).^(2/3);
    f_smooth   = @(x) exp(x);

    N_show = 80;
    x_show = cgl(N_show);
    a_naive = cheb_coeffs(f_full(x_show));
    a_trick = cheb_coeffs(f_smooth(x_show));

    Ns = [8 12 16 24 32 48 64 96 128 192 256];
    err_naive = zeros(size(Ns));
    err_trick = zeros(size(Ns));
    x_eval = linspace(-1.0, 1.0, 5001);
    for k = 1:numel(Ns)
        N = Ns(k);
        xg = cgl(N);
        a_n = cheb_coeffs(f_full(xg));
        a_t = cheb_coeffs(f_smooth(xg));
        approx_n = clenshaw(a_n, x_eval);
        approx_t = clenshaw(a_t, x_eval) + f_singular(x_eval);
        err_naive(k) = max(abs(approx_n - f_full(x_eval)));
        err_trick(k) = max(abs(approx_t - f_full(x_eval)));
    end

    fig = figure('Units', 'inches', 'Position', [1, 1, 11.5, 4.4], 'Color', 'w');
    tl = tiledlayout(1, 2, 'Padding', 'compact', 'TileSpacing', 'compact');

    nexttile(tl); hold on;
    n_axis = 0:N_show;
    semilogy(n_axis, abs(a_naive) + 1e-20, 'o-', 'Color', cm.CORAL, ...
             'MarkerFaceColor', 'w', 'MarkerSize', 4, 'LineWidth', 0.8, ...
             'DisplayName', 'naive: $|a_n|$ for $f(x)$');
    semilogy(n_axis, abs(a_trick) + 1e-20, '^-', 'Color', cm.TEAL, ...
             'MarkerFaceColor', 'w', 'MarkerSize', 4, 'LineWidth', 0.8, ...
             'DisplayName', 'trick: $|a_n|$ for $f - c(1+x)^{2/3}$');
    set(gca, 'YScale', 'log');
    hold off;
    xlabel('Chebyshev degree $n$', 'Interpreter', 'latex');
    ylabel('$|a_n|$', 'Interpreter', 'latex');
    title(sprintf('coefficient decay at $N = %d$', N_show), 'Interpreter', 'latex');
    ylim([1e-18, 5]);
    legend('Location', 'northeast', 'Interpreter', 'latex', 'FontSize', 9);
    grid on; box on;

    nexttile(tl); hold on;
    loglog(Ns, err_naive, 'o-', 'Color', cm.CORAL, 'MarkerFaceColor', 'w', ...
           'MarkerSize', 6, 'LineWidth', 1.0, 'DisplayName', 'naive: max $|f - f_N|$');
    loglog(Ns, err_trick, '^-', 'Color', cm.TEAL, 'MarkerFaceColor', 'w', ...
           'MarkerSize', 6, 'LineWidth', 1.0, ...
           'DisplayName', 'trick: max $|f - (\mathrm{smooth}_N + \mathrm{singular})|$');
    yline(1e-15, 'Color', [0.5 0.5 0.5], 'LineWidth', 0.4, 'Alpha', 0.5);
    set(gca, 'XScale', 'log', 'YScale', 'log');
    hold off;
    xlabel('$N$', 'Interpreter', 'latex');
    ylabel('max pointwise error', 'Interpreter', 'latex');
    title('subtraction reaches machine $\epsilon$ at modest $N$', 'Interpreter', 'latex');
    ylim([1e-17, 1]);
    legend('Location', 'southwest', 'Interpreter', 'latex', 'FontSize', 9);
    grid on; box on;

    exportgraphics(fig, fullfile(out_dir, 'singularity_subtraction.pdf'), 'ContentType', 'vector');
    exportgraphics(fig, fullfile(out_dir, 'singularity_subtraction.png'), 'Resolution', 300);
    close(fig);

    fprintf('[Etude 21.3]  singularity subtraction (1-D surrogate)\n');
    for k = 1:numel(Ns)
        fprintf('  N = %4d: naive err = %.3e, trick err = %.3e\n', ...
                Ns(k), err_naive(k), err_trick(k));
    end
    fprintf('  figure: %s\n', fullfile(out_dir, 'singularity_subtraction.pdf'));

    if ~isempty(dump_path)
        r = struct('Ns', Ns, 'err_naive', err_naive, 'err_trick', err_trick);
        fid = fopen(dump_path, 'w'); fprintf(fid, '%s', jsonencode(r)); fclose(fid);
    end
end

function dump_path = parse_args(varargin)
    dump_path = ''; k = 1;
    while k <= numel(varargin)
        if strcmp(varargin{k}, '--dump'); dump_path = char(varargin{k+1}); k = k + 2;
        else; k = k + 1; end
    end
end

function x = cgl(N)
    x = cos(pi * (0:N) / N);
end

function a = cheb_coeffs(v)
    N = numel(v) - 1;
    ext = [v(:); flipud(v(2:N).')]';
    A = real(fft(ext)) / N;
    A(1) = 0.5 * A(1);
    A(N+1) = 0.5 * A(N+1);
    a = A(1:N+1);
end

function val = clenshaw(a, t)
    N = numel(a) - 1;
    T0 = ones(size(t));
    if N == 0; val = a(1) * T0; return; end
    T1 = t;
    val = a(1) * T0 + a(2) * T1;
    for n = 2:N
        Tk = 2.0 * t .* T1 - T0;
        val = val + a(n+1) * Tk;
        T0 = T1; T1 = Tk;
    end
end

File: codes/julia/ch21/singularity_subtraction.jl
# singularity_subtraction.jl
# Chapter 21, Etude 21.3: singularity subtraction (1-D surrogate).
#
# f(x) = e^x + 0.1 (1+x)^(2/3) on [-1, 1].  Subtract the corner-like
# singular part; what remains is e^x, which has geometric Chebyshev
# convergence.
#
# Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)

using FFTW
include(joinpath(@__DIR__, "tricks_common.jl"))
apply_theme!()
const OUT = output_dir(@__DIR__)

const CSING = 0.1
f_full(x)     = exp(x) + CSING * (1.0 + x) ^ (2.0 / 3.0)
f_singular(x) = CSING * (1.0 + x) ^ (2.0 / 3.0)
f_smooth(x)   = exp(x)


cgl(N) = cos.(pi .* (0:N) ./ N)


function cheb_coeffs(v::AbstractVector)
    N = length(v) - 1
    ext = vcat(v, reverse(v[2:N]))
    A = real.(fft(ext)) ./ N
    A[1] *= 0.5; A[N+1] *= 0.5
    return A[1:N+1]
end


function reconstruct(a::AbstractVector, t_eval::AbstractVector)
    T0 = ones(length(t_eval))
    if length(a) == 1; return a[1] .* T0; end
    T1 = collect(t_eval)
    val = a[1] .* T0 .+ a[2] .* T1
    for n in 2:length(a)-1
        Tk = 2.0 .* t_eval .* T1 .- T0
        val .+= a[n+1] .* Tk
        T0, T1 = T1, Tk
    end
    return val
end


function make_figure(; dump_path = nothing)
    N_show = 80
    x_show = cgl(N_show)
    a_naive = cheb_coeffs(f_full.(x_show))
    a_trick = cheb_coeffs(f_smooth.(x_show))

    Ns = [8, 12, 16, 24, 32, 48, 64, 96, 128, 192, 256]
    err_naive = Float64[]
    err_trick = Float64[]
    x_eval = collect(range(-1.0, 1.0; length = 5001))
    for N in Ns
        xg = cgl(N)
        a_n = cheb_coeffs(f_full.(xg))
        a_t = cheb_coeffs(f_smooth.(xg))
        approx_n = reconstruct(a_n, x_eval)
        approx_t = reconstruct(a_t, x_eval) .+ f_singular.(x_eval)
        push!(err_naive, maximum(abs.(approx_n .- f_full.(x_eval))))
        push!(err_trick, maximum(abs.(approx_t .- f_full.(x_eval))))
    end

    fig = Figure(size = (1150, 440))

    ax1 = Axis(fig[1, 1]; xlabel = "Chebyshev degree n", ylabel = "|a_n|",
               yscale = log10,
               title = "coefficient decay at N = $N_show",
               limits = (nothing, (1e-18, 5.0)))
    n_axis = collect(0:N_show)
    scatterlines!(ax1, n_axis, abs.(a_naive) .+ 1e-20; color = CORAL,
        markercolor = :white, strokecolor = CORAL, strokewidth = 1.0,
        markersize = 4, linewidth = 0.8,
        label = "naive: |a_n| for f(x)")
    scatterlines!(ax1, n_axis, abs.(a_trick) .+ 1e-20; color = TEAL,
        marker = :utriangle, markercolor = :white, strokecolor = TEAL,
        strokewidth = 1.0, markersize = 4, linewidth = 0.8,
        label = "trick: |a_n| for f - c(1+x)^(2/3)")
    axislegend(ax1, position = :rt, labelsize = 9)

    ax2 = Axis(fig[1, 2]; xlabel = "N", ylabel = "max pointwise error",
               yscale = log10, xscale = log10,
               title = "subtraction reaches machine eps at modest N",
               limits = (nothing, (1e-17, 1.0)))
    scatterlines!(ax2, Ns, err_naive; color = CORAL, markercolor = :white,
        strokecolor = CORAL, strokewidth = 1.0, markersize = 6, linewidth = 1.0,
        label = "naive: max |f - f_N|")
    scatterlines!(ax2, Ns, err_trick; color = TEAL, marker = :utriangle,
        markercolor = :white, strokecolor = TEAL, strokewidth = 1.0,
        markersize = 6, linewidth = 1.0,
        label = "trick: max |f - (smooth_N + singular)|")
    hlines!(ax2, [1e-15]; color = :gray, linewidth = 0.4, alpha = 0.5)
    axislegend(ax2, position = :lb, labelsize = 9)

    save(joinpath(OUT, "singularity_subtraction.pdf"), fig)
    save(joinpath(OUT, "singularity_subtraction.png"), fig, px_per_unit = 4)

    println("[Etude 21.3]  singularity subtraction (1-D surrogate)")
    for (k, N) in enumerate(Ns)
        println("  N = $N:  naive err = $(err_naive[k]),  trick err = $(err_trick[k])")
    end
    println("  figure: ", joinpath(OUT, "singularity_subtraction.pdf"))

    if dump_path !== nothing
        write_json(dump_path, Dict("Ns" => Ns,
            "err_naive" => err_naive, "err_trick" => err_trick))
    end
end


function parse_args()
    dp = nothing; i = 1
    while i <= length(ARGS)
        if ARGS[i] == "--dump" && i < length(ARGS)
            dp = ARGS[i+1]; i += 2
        else; i += 1; end
    end
    return dp
end


if abspath(PROGRAM_FILE) == @__FILE__
    make_figure(; dump_path = parse_args())
end
